dds_web/database/models.py

```diff
@@ -174,45 +178,100 @@ class User(db.Model):
 
     @password.setter
     def password(self, plaintext_password):
         """Generate the password hash and save in db."""
         pw_hasher = argon2.PasswordHasher(hash_len=32)
+        self._password_hash = pw_hasher.hash(plaintext_password)
 
-        self._password = pw_hasher.hash(plaintext_password)
-
-    def verify_password_argon2id(self, input_password):
-        """Verifies that the password specified by the user matches
-        the encoded password in the database."""
-
+    def verify_password(self, input_password):
+        """Verifies that the specified password matches the encoded password in the database."""
         # Setup Argon2 hasher
         password_hasher = argon2.PasswordHasher(hash_len=32)
 
         # Verify the input password
         try:
-            password_hasher.verify(self.password, input_password)
+            password_hasher.verify(self._password_hash, input_password)
         except (
             argon2.exceptions.VerifyMismatchError,
             argon2.exceptions.VerificationError,
             argon2.exceptions.InvalidHash,
         ):
+            # Password hasher raises exceptions if not correct
             return False
 
-        # Rehash password if needed
-        if password_hasher.check_needs_rehash(self.password):
+        # Rehash password if needed, e.g. if parameters are not up to date
+        if not password_hasher.check_needs_rehash(self._password_hash):
             try:
-                self._password = password_hasher.hash(input_password)
+                self.password = input_password
                 db.session.commit()
             except sqlalchemy.exc.SQLAlchemyError as sqlerr:
                 db.session.rollback()
                 flask.current_app.logger.exception(sqlerr)
 
+        # Password correct
         return True
 
+    # 2FA related
+    def gen_otp_secret(self):
+        """Generate random base 32 for otp secret."""
+        return pyotp.random_base32()
+
+    @property
+    def otp_secret(self):
+        """Get OTP secret for user if the 2fa has not already been setup."""
+        if self.has_2fa:
+            return None
+        return self._otp_secret
+
+    @otp_secret.setter
+    def otp_secret(self, otp_secret):
+        """Set new otp secret."""
+        if not otp_secret:
+            otp_secret = self.gen_otp_secret()
+        self._otp_secret = otp_secret
+
+    def set_2fa_seen(self):
+        """Renew the otp secret -- one should only be setup once."""
+        self.has_2fa = True
+
+    def totp_uri(self):
+        """Get uri for user otp_secret if the 2fa has not already been setup."""
+        if self.has_2fa:
+            return None
+        return pyotp.totp.TOTP(self.otp_secret).provisioning_uri(
+            name=self.username, issuer_name="Data Delivery System"
+        )
+
+    def verify_totp(self, token):
+        """Verify the otp token."""
+        totp = pyotp.TOTP(self.otp_secret)
+        return totp.verify(token)
```

Comment:
I think there is an option to make this slightly more allowing (basically help for slow people like myself): `topt.verify(token, valid_window=1)`. This way it will be verified to True even though the timer has just flipped to the next interval.

Reply:
Oh, I didn't even think of that. Sure, sounds good!